File: DataImportManager/clsDataImportTask.vb
Imports PRISM.Logging
Imports System.Data.SqlClient
Imports System.IO
Imports DataImportManager.clsGlobal

Public Class clsDataImportTask
    Inherits clsDBTask

#Region "Member Variables"
    Private mPostTaskErrorMessage As String = String.Empty
    Private mDBErrorMessage As String

    Private mp_stored_proc As String
    Private mp_xmlContents As String
#End Region

#Region "Properties"

    Public ReadOnly Property PostTaskErrorMessage As String
        Get
            If String.IsNullOrEmpty(mPostTaskErrorMessage) Then
                Return String.Empty
            Else
                Return mPostTaskErrorMessage
            End If
        End Get
    End Property

    Public ReadOnly Property DBErrorMessage As String
        Get
            If String.IsNullOrEmpty(mDBErrorMessage) Then
                Return String.Empty
            Else
                Return mDBErrorMessage
            End If
        End Get
    End Property
    
    Public Property PreviewMode As Boolean

#End Region

    ''' <summary>
    ''' Constructor
    ''' </summary>
    ''' <param name="mgrParams"></param>
    ''' <param name="logger"></param>
    ''' <param name="dbConnection"></param>
    ''' <remarks></remarks>
    Public Sub New(mgrParams As IMgrParams, logger As ILogger, dbConnection As SqlConnection)
        MyBase.New(mgrParams, logger, dbConnection)
    End Sub

    Public Function PostTask(triggerFile As FileInfo) As Boolean

        mPostTaskErrorMessage = String.Empty
        mDBErrorMessage = String.Empty

        Dim fileImported As Boolean

        Try
            ' Call request stored procedure
            mp_xmlContents = LoadXmlFileContentsIntoString(triggerFile, m_logger)
            If String.IsNullOrEmpty(mp_xmlContents) Then
                Return False
            End If
            fileImported = ImportDataTask()
        Catch ex As Exception
            m_logger.PostEntry("clsDatasetImportTask.PostTask(), Error running PostTask, " & ex.Message, ILogger.logMsgType.logError, LOG_LOCAL_ONLY)
            Return False
        End Try

        Return fileImported

    End Function

    ''' <summary>
    ''' Posts the given XML to DMS5 using AddNewDataset
    ''' </summary>
    ''' <returns>True if success, false if an error</returns>
    ''' <remarks></remarks>
    Private Function ImportDataTask() As Boolean

        Dim Outcome As Boolean

        Try

            ' Initialize database error message
            mDBErrorMessage = String.Empty

            ' Prepare to call the stored procedure (typically AddNewDataset in DMS5, which in turn calls AddUpdateDataset)
            '
            mp_stored_proc = m_mgrParams.GetParam("storedprocedure")

            Dim sc = New SqlCommand(mp_stored_proc, m_DBCn)
            sc.CommandType = CommandType.StoredProcedure
            sc.CommandTimeout = 45

            '
            ' define parameter for stored procedure's return value
            '
            sc.Parameters.Add("@Return", SqlDbType.Int).Direction = ParameterDirection.ReturnValue
            sc.Parameters.Add("@XmlDoc", SqlDbType.VarChar, 4000).Value = mp_xmlContents
            sc.Parameters.Add("@mode", SqlDbType.VarChar, 24).Value = "add"
            sc.Parameters.Add("@message", SqlDbType.VarChar, 512).Direction = ParameterDirection.Output

            If PreviewMode Then
                ShowTraceMessage("Preview: call stored procedure " & mp_stored_proc & " in database " & m_DBCn.Database)
                Return True
            End If

            If TraceMode Then ShowTraceMessage("Calling stored procedure " & mp_stored_proc & " in database " & m_DBCn.Database)

            ' execute the stored procedure
            '
            sc.ExecuteNonQuery()

            ' get return value
            '
            Dim ret = CInt(sc.Parameters("@Return").Value)

            If ret = 0 Then
                ' get values for output parameters
                '
                Outcome = True
            Else
                mPostTaskErrorMessage = CStr(sc.Parameters("@message").Value)
                m_logger.PostEntry("clsDataImportTask.ImportDataTask(), Problem posting dataset: " & mPostTaskErrorMessage, ILogger.logMsgType.logError, LOG_LOCAL_ONLY)
                Outcome = False
            End If

        Catch ex As Exception
            m_logger.PostError("clsDataImportTask.ImportDataTask(), Error posting dataset: ", ex, True)
            mDBErrorMessage = ControlChars.NewLine & "Database Error Message:" & ex.Message
            Outcome = False
        End Try

        Return Outcome

    End Function

End Class

File: DataImportManager/clsDataImportTask.cs
﻿using System;
using System.Data;
using System.Text.RegularExpressions;
using PRISM.AppSettings;
using PRISMDatabaseUtils;

namespace DataImportManager
{
    // ReSharper disable once InconsistentNaming
    internal class clsDataImportTask : clsDBTask
    {
        #region "Member Variables"

        private string mPostTaskErrorMessage = string.Empty;
        private string mDatabaseErrorMessage;
        private string mStoredProc;
        private string mXmlContents;

        #endregion

        #region "Properties"

        public string PostTaskErrorMessage
        {
            get
            {
                if (string.IsNullOrEmpty(mPostTaskErrorMessage))
                {
                    return string.Empty;
                }

                return mPostTaskErrorMessage;
            }
        }

        public string DatabaseErrorMessage
        {
            get
            {
                if (string.IsNullOrEmpty(mDatabaseErrorMessage))
                {
                    return string.Empty;
                }

                return mDatabaseErrorMessage;
            }
        }

        /// <summary>
        /// When true, preview the datasets that would be added to DMS
        /// Also preview any e-mails that would be sent regarding errors
        /// </summary>
        public bool PreviewMode { get; set; }

        #endregion

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="mgrParams"></param>
        /// <param name="dbTools"></param>
        public clsDataImportTask(MgrSettings mgrParams, IDBTools dbTools) : base(mgrParams, dbTools)
        {
        }

        /// <summary>
        /// Send the contents of an XML trigger file to the database to create a new dataset
        /// </summary>
        /// <param name="triggerFileInfo"></param>
        public bool PostTask(TriggerFileInfo triggerFileInfo)
        {
            mPostTaskErrorMessage = string.Empty;
            mDatabaseErrorMessage = string.Empty;

            bool fileImported;
            try
            {
                // Load the XML file into memory
                mXmlContents = clsGlobal.LoadXmlFileContentsIntoString(triggerFileInfo.TriggerFile);
                if (string.IsNullOrEmpty(mXmlContents))
                {
                    return false;
                }

                // Check and modify contents if needed. Also report any replacements to the log file.
                if (triggerFileInfo.NeedsCaptureSubdirectoryReplacement)
                {
                    // <Parameter Name="Capture Subdirectory" Value="2020ESI_new.PRO\Data" />
                    var pattern = "(Parameter Name=\"Capture Sub(directory|folder)\" Value=\")" + Regex.Escape(triggerFileInfo.OriginalCaptureSubdirectory) + "\"";
                    mXmlContents = Regex.Replace(mXmlContents, pattern, $"$1{triggerFileInfo.FinalCaptureSubdirectory}\"", RegexOptions.IgnoreCase);
                    LogMessage($"Replaced capture subdirectory \"{triggerFileInfo.OriginalCaptureSubdirectory}\" with \"{triggerFileInfo.FinalCaptureSubdirectory}\"", writeToLog: true);
                }

                // Call the stored procedure (typically AddNewDataset)
                fileImported = ImportDataTask();
            }
            catch (Exception ex)
            {
                LogError("clsDatasetImportTask.PostTask(), Error running PostTask", ex);
                return false;
            }

            return fileImported;
        }

        /// <summary>
        /// Posts the given XML to DMS5 using stored procedure AddNewDataset
        /// </summary>
        /// <returns>True if success, false if an error</returns>
        private bool ImportDataTask()
        {
            try
            {
                // Initialize database error message
                mDatabaseErrorMessage = string.Empty;

                // Prepare to call the stored procedure (typically AddNewDataset in DMS5, which in turn calls AddUpdateDataset)
                mStoredProc = MgrParams.GetParam("StoredProcedure");

                var cmd = DBTools.CreateCommand(mStoredProc, CommandType.StoredProcedure);
                cmd.CommandTimeout = 45;

                // Define parameter for stored procedure's return value
                var returnParam = DBTools.AddParameter(cmd, "@Return", SqlType.Int, direction: ParameterDirection.ReturnValue);
                DBTools.AddParameter(cmd, "@XmlDoc", SqlType.VarChar, 4000, mXmlContents);
                DBTools.AddParameter(cmd, "@mode", SqlType.VarChar, 24, "add");
                var messageParam = DBTools.AddParameter(cmd, "@message", SqlType.VarChar, 512, direction: ParameterDirection.Output);

                if (PreviewMode)
                {
                    clsMainProcess.ShowTraceMessage("Preview: call stored procedure " + mStoredProc + " in database " + DBTools.DatabaseName);
                    return true;
                }

                if (TraceMode)
                {
                    clsMainProcess.ShowTraceMessage("Calling stored procedure " + mStoredProc + " in database " + DBTools.DatabaseName);
                }

                // Execute the stored procedure
                var returnCode = DBTools.ExecuteSP(cmd);

                // Get return code
                var returnParamText = DBTools.GetString(returnParam.Value);

                if (returnCode == 0)
                {
                    // Get values for output parameters
                    return true;
                }

                mPostTaskErrorMessage = DBTools.GetString(messageParam.Value) ?? string.Empty;
                LogError(string.Format("clsDataImportTask.ImportDataTask(), Problem posting dataset (return code {0}): {1}",
                    returnParamText ?? string.Empty, mPostTaskErrorMessage));

                return false;
            }
            catch (Exception ex)
            {
                LogError("clsDataImportTask.ImportDataTask(), Error posting dataset", ex, true);
                mDatabaseErrorMessage = Environment.NewLine + ("Database Error Message:" + ex.Message);
                return false;
            }
        }
    }
}
